x/storage/keeper/payment.go

```diff
@@ -100,218 +148,353 @@ func (k Keeper) UnlockStoreFee(ctx sdk.Context, bucketInfo *storagetypes.BucketI
 }
 
-func (k Keeper) ChargeStoreFee(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo, objectInfo *storagetypes.ObjectInfo) error {
-	chargeSize, err := k.GetChargeSize(ctx, objectInfo.PayloadSize, objectInfo.CreateAt)
+func (k Keeper) ChargeObjectStoreFee(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo,
+	internalBucketInfo *storagetypes.InternalBucketInfo, objectInfo *storagetypes.ObjectInfo) error {
+	chargeSize, err := k.GetObjectChargeSize(ctx, objectInfo.PayloadSize, objectInfo.CreateAt)
 	if err != nil {
 		return fmt.Errorf("get charge size error: %w", err)
 	}
-	return k.ChargeViaBucketChange(ctx, bucketInfo, func(bi *storagetypes.BucketInfo) error {
-		bi.BillingInfo.TotalChargeSize += chargeSize
-		secondarySpObjectsSize := bi.BillingInfo.SecondarySpObjectsSize
-		for _, sp := range objectInfo.SecondarySpAddresses {
-			secondarySpObjectsSize = append(secondarySpObjectsSize, storagetypes.SecondarySpObjectsSize{
-				SpAddress:       sp,
-				TotalChargeSize: chargeSize,
-			})
+
+	priceChanged, err := k.IsPriceChanged(ctx, bucketInfo.PrimarySpId, internalBucketInfo.PriceTime)
+	if err != nil {
+		return fmt.Errorf("check whether price changed error: %w", err)
+	}
+
+	if !priceChanged {
+		err := k.ChargeViaObjectChange(ctx, bucketInfo, internalBucketInfo, objectInfo, chargeSize, false)
+		if err != nil {
+			return fmt.Errorf("apply object store bill error: %w", err)
+		}
+		return nil
+	}
+
+	return k.ChargeViaBucketChange(ctx, bucketInfo, internalBucketInfo, func(bi *storagetypes.BucketInfo, ibi *storagetypes.InternalBucketInfo) error {
+		ibi.TotalChargeSize += chargeSize
+		for _, lvg := range ibi.LocalVirtualGroups {
+			lvg.TotalChargeSize += lvg.TotalChargeSize + chargeSize
+		}
+		return nil
+	})
+}
+
+func (k Keeper) UnChargeObjectStoreFee(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo,
+	internalBucketInfo *storagetypes.InternalBucketInfo, objectInfo *storagetypes.ObjectInfo) error {
+	chargeSize, err := k.GetObjectChargeSize(ctx, objectInfo.PayloadSize, objectInfo.CreateAt)
+	if err != nil {
+		return fmt.Errorf("get charge size error: %w", err)
+	}
+
+	priceChanged, err := k.IsPriceChanged(ctx, bucketInfo.PrimarySpId, internalBucketInfo.PriceTime)
+	if err != nil {
+		return fmt.Errorf("check whether price changed error: %w", err)
+	}
+
+	if !priceChanged {
+		err := k.ChargeViaObjectChange(ctx, bucketInfo, internalBucketInfo, objectInfo, chargeSize, true)
+		if err != nil {
+			return fmt.Errorf("apply object store bill error: %w", err)
+		}
+		return nil
+	}
+
+	return k.ChargeViaBucketChange(ctx, bucketInfo, internalBucketInfo, func(bi *storagetypes.BucketInfo, ibi *storagetypes.InternalBucketInfo) error {
+		ibi.TotalChargeSize -= chargeSize
+		lvgs := make([]*storagetypes.LocalVirtualGroup, 0)
+		for _, lvg := range ibi.LocalVirtualGroups {
+			if lvg.Id == objectInfo.LocalVirtualGroupId {
+				lvg.TotalChargeSize -= chargeSize
+			}
+			lvgs = append(lvgs, lvg)
```

Comment:
Do we really need another `lvgs` here, I think we can assign in place.

Reply:
no need create new lvgs. fixed.